rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,482 @@
+trigger for dynamic node creation will be incoming invocations. The label of the
+channel associated with the request receiver on the invocation will be used to
+match the template label pattern and extract the parameters.
+
+A sub-graph template will also be able to define whether it is persistent or
+not. If it is persistent, an instance will be created the first time it is
+needed for a specific label. For future instances of the same label it will
+reuse the already-created nodes and channels. If it is not persistent, the
+subgraph will be recreated every time a matching label arrives. It is the
+responsibility of the nodes in question to close the channels and terminate once
+processing is complete to avoid running out of resources.
+
+The template will contain definitions of nodes and channels to be created. These
+definitions can use the parameters for specifying labels, names etc. The
+template will support a formatting mechanism to replace placeholders with
+parameter values. In most cases the parameter values will be byte arrays. The
+formatting will allow specification of conversions to strings so that it could
+be used in names etc. Initially only base64 encoding will be supported for
+converting from bytes to strings.
+
+As an example, if a parameter named `user_id` is extracted for the label, a
+user-specific node can use the base64 encoding of the user id to create a unique
+name:
+
+```toml
+[node_template]
+name = "user_{base64($user_id)}"
+
+```
+
+The initial impementation of dynamic node creation will focus only on support
+for per-user labels, as it is the first motivating example we have. Each
+extracted parameter will be the matching bytes that represent the user identity
+(currently the HMAC of the bearer token, but in future it would be the user's
+public key).
+
+A new LabelTemplate protobuf object will be implemented to support parameterised
+labels. It will reuse the existing tags from the Label, but also add templated
+tags. These can be used for matching labels and extracting parameters, as well
+as generating concrete label instances given a parameter value.
+
+```proto
+
+// Template for generating labels given parameter values, or matching labels
+// and extracting parameter values.
+message LabelTemplate {
+  repeated TagTemplate confidentiality_tags = 1;
+  repeated TagTemplate integrity_tags = 2;
+}
+
+// Template for generating a user-specific gRPC tag given a user ID, or for
+// extracting a user ID parameter from a matched gRPC tag.
+message GrpcTagTemplate {
+  string user_id_parameter_name = 1;
+}
+
+
+// Templated tags for supporting templated labels.
+message TagTemplate {
+  oneof tag {
+    GrpcTag grpc_tag = 1;
+    GrpcTagTemplate grpc_tag_template = 1;
+    WebAssemblyModuleTag web_assembly_module_tag = 2;
+    WebAssemblyModuleSignatureTag web_assembly_module_signature_tag = 3;
+    TlsEndpointTag tls_endpoint_tag = 4;
+  }
+}
```

Comment:
Is this hierarchy identical to the label hierarchy? If yes, why do we need it? If not, could you highlight what is different?

Reply:
The GrpcTagTemplate is extra. It is a parameterised version of the user-specific GrpcTag.

Reply:
Could you add this explanation to the doc?